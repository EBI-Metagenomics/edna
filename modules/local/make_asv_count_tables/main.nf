
process MAKE_ASV_COUNT_TABLES {
    tag "$meta.id"
    label 'medium'
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        "https://depot.galaxyproject.org/singularity/mgnify-pipelines-toolkit:${params.mpt_version}":
        "biocontainers/mgnify-pipelines-toolkit:${params.mpt_version}" }"

    input:
    tuple val(meta), path(maps), path(asvtaxtable), path(reads), path(extracted_var_path)
    val db_label

    output:
    tuple val(meta), path("*asv_krona_counts.txt"), optional: true, emit: asv_count_tables_out
    tuple val(meta), path("*asv_read_counts.tsv") , optional: true, emit: asv_read_counts_out
    tuple val(meta), path("*asvs_left.txt")       , optional: true, emit: asvs_left
    path "versions.yml"                           , emit: versions

    script:
    """
    if [[ ${meta.single_end} = true ]]; then
        zcat $reads | awk 'NR % 4 == 1' > headers.txt
        make_asv_count_table -t $asvtaxtable -f $maps -a $extracted_var_path -hd ./headers.txt  -s ${meta.id}_${meta.var_region}_${db_label}
        tail -n +2 *_asv_read_counts.tsv | cut -f1 > ${meta.id}_${meta.var_region}_${db_label}_asvs_left.txt
    else
        zcat ${reads[0]} | awk 'NR % 4 == 1' > headers.txt
        make_asv_count_table -t $asvtaxtable -f ${maps[0]} -r ${maps[1]} -a $extracted_var_path -hd ./headers.txt  -s ${meta.id}_${meta.var_region}_${db_label}
        tail -n +2 *_asv_read_counts.tsv | cut -f1 > ${meta.id}_${meta.var_region}_${db_label}_asvs_left.txt
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        mgnify-pipelines-toolkit: ${params.mpt_version}
    END_VERSIONS
    """
}
