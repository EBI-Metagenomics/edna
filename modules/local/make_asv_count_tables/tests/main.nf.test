nextflow_process {

    name "Test Process MAKE_ASV_COUNT_TABLES"
    script "../main.nf"
    process "MAKE_ASV_COUNT_TABLES"
    tag "modules"
    tag "make_asv_count_tables"

    test("Should run without failures") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false, var_region:'16S-V3-V4' ], // meta map
                    [
                        file('${projectDir}/modules/local/make_asv_count_tables/tests/data/test_1_map.txt', checkIfExists: true),
                        file('${projectDir}/modules/local/make_asv_count_tables/tests/data/test_2_map.txt', checkIfExists: true) 
                    ],
                    file('${projectDir}/modules/local/make_asv_count_tables/tests/data/test_taxa.tsv', checkIfExists: true),
                    [ 
                        file('${projectDir}/modules/local/make_asv_count_tables/tests/data/test_1.fastp.fastq.gz', checkIfExists: true),
                        file('${projectDir}/modules/local/make_asv_count_tables/tests/data/test_2.fastp.fastq.gz', checkIfExists: true) 
                    ],
                    file('${projectDir}/modules/local/make_asv_count_tables/tests/data/test.16S.V3-V4.txt', checkIfExists: true)
                ]
                input[1] = "DADA2-PR2"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.asv_count_tables_out).match("asv_count_tables_out") }
            )
        }

    }

}
