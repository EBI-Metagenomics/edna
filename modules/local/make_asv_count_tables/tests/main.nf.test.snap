{
    "asv_count_tables_out": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false,
                        "var_region": "16S-V3-V4"
                    },
                    "test_16S-V3-V4_DADA2-PR2_16S.V3-V4_pr2_asv_krona_counts.txt:md5,ec5d39eef4d33386346c3f6fa007b59d"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-24T16:19:29.356317047"
    }
}